simple_network_sim/sns_logger.py

```diff
@@ -0,0 +1,108 @@
+should - for consistency - be done before any information is written
+to the logger.
+
+To use the package-level logger, you might use
+
+from . import sns_logger
+logger = sns_logger.logger
+
+or
+
+from simple_network_sim import sns_logger
+logger = sns_logger.logger
+
+NOTE: By default, the logger writes to sys.stderr and only reports
+logger.WARNING and above.
+"""
+
+import logging
+import logging.config
+import sys
+
+from argparse import Namespace
+from pathlib import Path
+from typing import Optional
+
+def setup_logger(args: Optional[Namespace] = None) -> None:
+    """Create and return a logging.Logger instance.
+    
+    :param args: argparse.Namespace
+        args.logfile (pathlib.Path) is used to create a logfile if present
+        args.verbose and args.debug control logging level to sys.stderr
+    """
+    # Dictionary to define logging configuration
+    logconf = {
+        "version": 1,
+        "disable_existing_loggers": False,
+        "formatters": {
+            "standard": {
+                "format": "[%(asctime)s] [%(name)s] [%(levelname)s] %(message)s",
+                },
+            },
+        "handlers": {
+            "stderr": {
+                "level": "WARNING",
+                "class": "logging.StreamHandler",
+                "formatter": "standard",
+                "stream": "ext://sys.stderr"
+                },
+            },
+        "loggers": {
+            __package__: {
+                "handlers": ["stderr"],
+                "level": "DEBUG",
+                },
+            },
+        }
+
+    # If args.logpath is specified, add logfile
+    if args is not None and args.logfile is not None:
+        logdir = args.logfile.parents[0]
+        # If the logfile is going in another directory, we must
+        # create it/check if it's there
+        try:
+            if not logdir == Path.cwd():
+                logdir.mkdir(exist_ok=True)
+        except OSError:
+            logger.error("Could not create %s for logging", logdir, exc_info=True)
+            raise SystemExit(1)  # Substitute meaningful error code when known
+        # Add logfile configuration
+        logconf["handlers"]["logfile"] = {
+            "class": "logging.FileHandler",
+            "level": "INFO",
+            "formatter": "standard",
+            "filename": str(args.logfile),
+            "encoding": "utf8",
+            }
+        logconf["loggers"][__package__]["handlers"].append("logfile")
+
+    # Set STDERR/logfile levels if args.verbose/args.debug specified
+    if args is not None and args.verbose:
+        logconf["handlers"]["stderr"]["level"] = "INFO"
+    elif args is not None and args.debug:
+        logconf["handlers"]["stderr"]["level"] = "DEBUG"
+        logconf["handlers"]["logfile"]["level"] = "DEBUG"
+
+    # Configure logger
+    logging.config.dictConfig(logconf)
+
+# Provide a package-level logging.Logger object on import
+logger = logging.getLogger(__package__)
```

Comment:
By not calling this inside each module we won't have the module name in the log line anymore. I don't see what's the issue we are preventing by creating this here.

Reply:
It wasn't clear to me that having the module name in the log line was a requirement.

Calling `logging.getLogger(__name__)` creates a new logger with its own scope, that also requires its own configuration. This brings extra overhead in coding, but doesn't give extra functionality. This makes adding extra handlers more difficult, too.

All modules in a package are related at the package level, and a common logger needs only to be configured once to write to the appropriate stream(s) (e.g. a syslog server, log file, STDOUT, etc.). All modules contribute automatically to the same stream(s), which may be set to distinct reporting levels.

If there is a need to identify the module name for information/help with debuggin, this can still be done explicitly in the log message, e.g. `logger.info("Failed in module %s.", __name__), or by including the exception info with `exc_info=True`).

Reply:
We don't need to configure each logger individualy, we can just configure the root logger. All the other loggers will inherit from it. By calling `logging.getLogger(__name__)` in each module we get hierarchical logging for free. Python automatically parses `a.b.c` to create the hierarchical logging. This has benefits like changing the log level of each package or module individually if we wanted (and also the module name in the logging line, which I find nice). I think the main reason I'm questioning it is that using `logging.getLogger` is sort of the recommended way to do it in the python docs, I was just trying to understand better the issue with it. Sorry if I'm being annoying, I don't think this is super important.

Reply:
Don't worry - it's not annoying. It's helping me rationalise and understand things better (but it's getting late).